pkg/apis/pipeline/v1/pipeline_types.go

```diff
@@ -0,0 +1,566 @@
+	// Params declares a list of input parameters that must be supplied when
+	// this Pipeline is run.
+	// +listType=atomic
+	Params []ParamSpec `json:"params,omitempty"`
+	// Workspaces declares a set of named workspaces that are expected to be
+	// provided by a PipelineRun.
+	// +optional
+	// +listType=atomic
+	Workspaces []PipelineWorkspaceDeclaration `json:"workspaces,omitempty"`
+	// Results are values that this pipeline can output once run
+	// +optional
+	// +listType=atomic
+	Results []PipelineResult `json:"results,omitempty"`
+	// Finally declares the list of Tasks that execute just before leaving the Pipeline
+	// i.e. either after all Tasks are finished executing successfully
+	// or after a failure which would result in ending the Pipeline
+	// +listType=atomic
+	Finally []PipelineTask `json:"finally,omitempty"`
+}
+
+// PipelineResult used to describe the results of a pipeline
+type PipelineResult struct {
+	// Name the given name
+	Name string `json:"name"`
+
+	// Type is the user-specified type of the result.
+	// The possible types are 'string', 'array', and 'object', with 'string' as the default.
+	// 'array' and 'object' types are alpha features.
+	Type ResultsType `json:"type,omitempty"`
+
+	// Description is a human-readable description of the result
+	// +optional
+	Description string `json:"description"`
+
+	// Value the expression used to retrieve the value
+	Value ArrayOrString `json:"value"`
+}
+
+// PipelineTaskMetadata contains the labels or annotations for an EmbeddedTask
+type PipelineTaskMetadata struct {
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+// EmbeddedTask is used to define a Task inline within a Pipeline's PipelineTasks.
+type EmbeddedTask struct {
+	// +optional
+	runtime.TypeMeta `json:",inline,omitempty"`
+
+	// Spec is a specification of a custom task
+	// +optional
+	Spec runtime.RawExtension `json:"spec,omitempty"`
+
+	// +optional
+	Metadata PipelineTaskMetadata `json:"metadata,omitempty"`
+
+	// TaskSpec is a specification of a task
+	// +optional
+	TaskSpec `json:",inline,omitempty"`
+}
+
+// PipelineTask defines a task in a Pipeline, passing inputs from both
+// Params and from the output of previous tasks.
+type PipelineTask struct {
+	// Name is the name of this task within the context of a Pipeline. Name is
+	// used as a coordinate with the `from` and `runAfter` fields to establish
+	// the execution order of tasks relative to one another.
+	Name string `json:"name,omitempty"`
+
+	// TaskRef is a reference to a task definition.
+	// +optional
+	TaskRef *TaskRef `json:"taskRef,omitempty"`
+
+	// TaskSpec is a specification of a task
+	// +optional
+	TaskSpec *EmbeddedTask `json:"taskSpec,omitempty"`
+
+	// WhenExpressions is a list of when expressions that need to be true for the task to run
+	// +optional
+	WhenExpressions WhenExpressions `json:"whenExpressions,omitempty"`
```

Comment:
Please don't change the json tag here; this changes the API which there's not a good reason to do here. Instead, change the name of the struct field to "when"

Reply:
Thanks Lee. In order to show it's a list, would it make any sense to rename it as `Whens`?

Reply:
What you have now is correct! It looks like you just need to re-run codegen/openapigen.

Reply:
Thanks @lbernick. It always got me that the `update-codegen` won't trigger `openapigen`, but it should.